pkg/cloudprovider/aws/launchtemplate.go

```diff
@@ -19,139 +19,137 @@ import (
+		amiProvider:           amiProvider,
+		securityGroupProvider: securityGroupProvider,
+		cache:                 cache.New(CacheTTL, CacheCleanupInterval),
+	}
 }
 
 func launchTemplateName(options *launchTemplateOptions) string {
 	hash, err := hashstructure.Hash(options, hashstructure.FormatV2, nil)
 	if err != nil {
 		zap.S().Panicf("hashing launch template, %w", err)
 	}
-	return fmt.Sprintf(launchTemplateNameFormat, options.Cluster.Name, options.Provisioner.Name, options.Provisioner.Namespace, fmt.Sprint(hash))
+	return fmt.Sprintf(launchTemplateNameFormat, options.Cluster.Name, fmt.Sprint(hash))
 }
 
 // launchTemplateOptions is hashed and results in the creation of a real EC2
 // LaunchTemplate. Do not change this struct without thinking through the impact
 // to the number of LaunchTemplates that will result from this change.
 type launchTemplateOptions struct {
-	Provisioner  types.NamespacedName
-	Cluster      v1alpha1.ClusterSpec
-	Architecture string
-	Labels       map[string]string
-	Taints       []v1.Taint
+	// Edge-triggered fields that will only change on kube events.
+	Cluster  v1alpha1.ClusterSpec
+	UserData string
+	// Level-triggered fields that may change out of sync.
+	SecurityGroups []string
+	AMIID          string
 }
 
 func (p *LaunchTemplateProvider) Get(ctx context.Context, provisioner *v1alpha1.Provisioner, constraints *Constraints) (*LaunchTemplate, error) {
-	// If the customer specified a launch template then just use it
+	// 1. If the customer specified a launch template then just use it
 	if result := constraints.GetLaunchTemplate(); result != nil {
 		return result, nil
 	}
 
-	options := launchTemplateOptions{
-		Provisioner:  types.NamespacedName{Name: provisioner.Name, Namespace: provisioner.Namespace},
-		Cluster:      *provisioner.Spec.Cluster,
-		Architecture: KubeToAWSArchitectures[*constraints.Architecture],
-		Labels:       constraints.Labels,
-		Taints:       constraints.Taints,
-	}
-	// See if we have a cached copy of the default one first, to avoid
-	// making an API call to EC2
-	key, err := hashstructure.Hash(options, hashstructure.FormatV2, nil)
+	// 2. Get constrained AMI ID
+	amiID, err := p.amiProvider.Get(ctx, constraints)
 	if err != nil {
-		return nil, fmt.Errorf("hashing launch template, %w", err)
+		return nil, err
 	}
 
-	result := &LaunchTemplate{Version: aws.String(DefaultLaunchTemplateVersion)}
-	if cached, ok := p.cache.Get(fmt.Sprint(key)); ok {
-		result.Id = cached.(*ec2.LaunchTemplate).LaunchTemplateId
-		return result, nil
+	// 3. Get constrained security groups
+	securityGroups, err := p.getSecurityGroupIds(ctx, provisioner, constraints)
+	if err != nil {
+		return nil, err
 	}
 
-	// Call EC2 to get launch template, creating if necessary
-	launchTemplate, err := p.getLaunchTemplate(ctx, &options)
+	// 4. Ensure the launch template exists, or create it
+	launchTemplate, err := p.ensureLaunchTemplate(ctx, &launchTemplateOptions{
+		Cluster:        *provisioner.Spec.Cluster,
```

Comment:
It looks like the launch templates are pulled from the spec of the provisioner. Just wanted to check that we can't create different Launch Templates for the same cluster from different provisioners? Seems like based on this PR's changes, we cannot.

Reply:
The cache key is the name, which is a hash of launchTemplateOptions. You can have many launch templates per cluster.